File: lambdas/serviciocliente/CrearReview.py
import boto3
import json
from datetime import datetime

def lambda_handler(event, context):
    dynamodb = boto3.resource('dynamodb')
    table = dynamodb.Table('t_reviews')

    tenant_id = event['body']['tenant_id']
    user_id = event['body']['user_id']
    product_id = event['body'].get('product_id')  # Optional for general questions
    order_id = event['body'].get('order_id')      # Optional for order reviews
    question_text = event['body']['question_text']
    review_type = event['body']['review_type']  # e.g., 'product', 'order', 'general'
    timestamp = datetime.utcnow().isoformat()

    if not (tenant_id and user_id and question_text and review_type):
        return {
            'statusCode': 400,
            'body': json.dumps({'error': 'Invalid request body'})
        }

    table.put_item(
        Item={
            'tenant_id': tenant_id,
            'user_id': user_id,
            'product_id': product_id,
            'order_id': order_id,
            'question_text': question_text,
            'review_type': review_type,
            'timestamp': timestamp,
            'answer': '',
            'answertimestamp': ''
        }
    )

    return {
        'statusCode': 201,
        'body': json.dumps({
            'message': 'Review/Question created successfully',
            'tenant_id': tenant_id,
            'user_id': user_id,
            'review_type': review_type,
            'timestamp': timestamp
        })
    }

File: lambdas/comentario/CrearReview.js
const { DynamoDBClient } = require("@aws-sdk/client-dynamodb");
const { DynamoDBDocumentClient, PutCommand } = require("@aws-sdk/lib-dynamodb");

const client = new DynamoDBClient({});
const dynamo = DynamoDBDocumentClient.from(client);

const https = require('https');

const validateToken = (token) => {
  return new Promise((resolve, reject) => {
    const options = {
      hostname: 'bfh1meojk2.execute-api.us-east-1.amazonaws.com',
      path: '/prod/token/validate',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`, 
      },
    };

    const req = https.request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => {
        data += chunk;
      });
      res.on('end', () => {
        const response = JSON.parse(data);
        if (res.statusCode === 200) {
          resolve(response); 
        } else {
          reject(new Error(response.error || 'Token no válido')); 
        }
      });
    });

    req.on('error', (e) => {
      reject(new Error(`Error en la validación del token: ${e.message}`));
    });

    req.end();
  });
};

exports.lambda_handler = async (event) => {
  try {
    const token = event.headers.Authorization?.split(' ')[1]; 
    if (!token) {
      return {
        statusCode: 400,
        body: { message: "Token is required" },
      };
    }

    try {
      await validateToken(token); 
    } catch (error) {
      return {
        statusCode: 403,
        body: { message: error.message },
      };
    }

    const body = typeof event.body === "string" ? JSON.parse(event.body) : event.body;
    const {tenant_id, product_id, review_id, user_id, comentario, stars} = body

    const lastModification = new Date().toISOString();

    await dynamo.send(
      new PutCommand({
        TableName: "pf_comentario",
        Item: {
          tenant_id: tenant_id,
          pr_id: `${product_id}#${review_id}`,
          product_id: product_id,
          review_id: review_id,
          user_id: user_id,
          comentario: comentario,
          stars: Number(stars),
          last_modification:lastModification
        },
      })
    );

    return {
      statusCode: 201,
      body: {
        message: "review created successfully",
      },
    };
  } catch (error) {
    return {
      statusCode: 500,
      body: {
        error: error.message || "An error occurred while creating the review",
      },
    };
  }
};
